### DISCUSSION TITLE
A Freestanding Rust Binary (zh-CN)

### DISCUSSION THREAD

**Discussion by giscus** [phil-opp/blog_os#1184] [Post Comments (translated)]
# A Freestanding Rust Binary (zh-CN)

创建一个不链接标准库的 Rust 可执行文件，将是我们迈出的第一步。无需底层操作系统的支撑，这样才能在裸机（bare metal）上运行 Rust 代码。

https://os.phil-opp.com/zh-CN/freestanding-rust-binary/

**Comment by pengwon** [comment]
great job!

**Comment by dreamerlzl** [comment]
I also recommend reading the Chapter "Rust without the standard library" in *Rust for Rustaceans* by Jon as a complement

**Reply by pengwon** [reply]
thanks

**Reply by dftdo** [reply]
assurance

**Comment by Nazaroni** [comment]
It looks promising!

**Comment by dftdo** [comment]
good job!!
I also recommend you to take a look at creating an OS in assembly and c to understand a little more

**Reply by ashupadhi01** [reply]
Have you done that ? If yes, which resources you followed ? Give me some idea.

**Comment by YoungHaKim7** [comment]
thank you 👍

**Comment by cshuo** [comment]
looks good.

**Comment by KeplerIO** [comment]
There is a Tier 2 compilation target triple: "x86_64-unknown-none"

Looks like it is the perfect use case for this example. 

https://doc.rust-lang.org/nightly/rustc/platform-support/x86_64-unknown-none.html

**Reply by phil-opp** [reply]
Yes, we will use that target for the third edition of the blog. You can read the current draft at https://github.com/phil-opp/blog_os/blob/edition-3/blog/content/edition-3/posts/01-minimal-kernel/index.md

**Comment by FelixSelter** [comment]
Thank you for the tutorial. My panic! handler was not working though. The rust compiler would break the entire code when I call panic!. That`s because there`s still some minimal unwinding which uses the stackpointer. As the stack was not setup properly the kernel tried to access negative memory.

To see how to get your panic handler working refer to this discussion: https://users.rust-lang.org/t/bare-metal-panic-generates-undefined-instructions/94992/9

**Reply by phil-opp** [reply]
Thanks for reporting! I didn't see this error before. Do you have your code online somewhere so that I can debug?

**Reply by FelixSelter** [reply]
I can publish it. Its very simple

**Reply by phil-opp** [reply]
That would be great!

**Comment by Lisheri** [comment]
nb。 It's too strong

**Comment by 220303** [comment]
Windows平台上项目目录不要太长，否则编译过程中链接器(rust-lld)会报错：
```
error: linking with `rust-lld` failed: exit code: 0xc0000374
  |
  = note: "rust-lld" "-flavor" "gnu" "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\rustci2Ef3q\\symbols.o" "<1 object files omitted>" "--as-needed" "-Bstatic" "T:\\ELCO\\Computer Science\\Operating System\\220303\'s Kernel R\\kernel\\target\\x86_64-kernel\\debug\\deps/{librustc_std_workspace_core-e5e018c30b9b1f48.rlib,libcore-8c663417353e0152.rlib,libcompiler_builtins-918324d46c04c70b.rlib}.rlib" "-L" "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\rustci2Ef3q\\raw-dylibs" "-Bdynamic" "--eh-frame-hdr" "-z" "noexecstack" "-o" "T:\\ELCO\\Computer Science\\Operating System\\220303\'s Kernel R\\kernel\\target\\x86_64-kernel\\debug\\deps\\kernel-c8d1dc174c6d25a5" "--gc-sections"        
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note:
```
我原来的目录就是因为太长所以报错，直接把项目目录剪切桌面上就好了，文件目录里有空格和特殊字符(我用了英文 ’ 字符)，目前没有什么问题。

用Vs code可能会在main.rs中 #![no_std] 那一行提示 can't find crate for test ，但只要你自己确定创建包的时候时bin而不是lib (默认不加参数是bin)，就不用管这个报错，不影响编译。
